### ISSUE TITLE
Migrate to lucene 6

### ISSUE THREAD

**Issue by chaitanya9186** [cloudant-labs/clouseau#4]
Lucene 6 requires jdk 8 but clouseau depends on scalang which needs jdk 6.

So
1. We need to wait for Scala to support jdk 8, but full java 8 support is in Scala 2.12's [road map](http://scala-lang.org/news/2.12-roadmap).
2. Migrate [scalang](https://github.com/boundary/scalang) to jdk 8.
3. Migrate [clouseau](https://github.com/cloudant-labs/clouseau) to jdk 8.
4. Migrate [clouseau](https://github.com/cloudant-labs/clouseau) to lucene 6.

**Comment by dmunch** [comment]
Good news! Scala 2.12 seems to be ready, yesterday they tagged 2.12.0-RC2 as 2.12.0: https://github.com/scala/scala/releases/tag/v2.12.0

To make things move forward and take advantage of that grey and cloudy Berlin automn weather I ported overlock to Scala 2.12 and JDK8. overlock is a dependency of scalang and was the only dependency not yet available on JDK8. 

The ported version is available here:
https://github.com/dmunch/overlock/tree/scala-2120

**Comment by chaitanya9186** [comment]
Nice to hear that.

**Comment by rnewson** [comment]
Cool!

**Comment by dmunch** [comment]
Just pushed the corresponding changes to scalang and clouseau and I'm happy to announce that all tests pass and that clouseau seems to be happily running on Java 8 now! 
- The packages still use Scala 2.12.0-RC2. It's identical to Scala 2.12.0, it's just that as far as I could see all dependencies didn't upgrade to 2.12.0 yet.
- Some deprecation warnings are left which might be good to tackle. 
- Scalang tests are still a little flaky, I usually need to run them two or three times in order for them to pass.
- As an intermediary solution I used bintray to host the poms which doesn't allow to push snapshots. 
- I did ask for a JCenter account for scalang and overlock which might be nice for a setting up a CI pipeline, not sure if it's going to be granted though. 

Anyway, here are the sources: 

https://github.com/dmunch/overlock/tree/scala-2120
https://github.com/dmunch/scalang/tree/scala-2120
https://github.com/dmunch/clouseau/tree/scala-2120

**Comment by rnewson** [comment]
Great progress! I can post the binaries up to maven.cloudant.com in the week as a more permanent home.

**Comment by rnewson** [comment]
Just a bump. Not likely to get to work on this till next week but it's not forgotten.

**Comment by dmunch** [comment]
No worries, there's no hurry from my side! 

I started working on porting to Lucene 6.3.0 and I'm somewhere halfway through. It compiles and most test run and succeed (minus those I disabled for the moment..), however the parts concerning faceting are all commented out. 

The WIP work can be found here: https://github.com/dmunch/clouseau/commits/scala-2120

**Comment by dmunch** [comment]
Another rainy Saturday: Facets are working again! 

However, in Lucene 6 apparently there has to be a differentiation between DoubleRangeFacetCounts and LongRangeFacetCounts.

**Comment by rnewson** [comment]
For simplicity, I'd forced all numbers to be doubles anyway. That's something to maybe reconsider but given we pass through a javascript engine (spidermonkey) where all numbers are floats, I'm not sure there's much point.

**Comment by dmunch** [comment]
Yeah I realized that quickly after. I was just getting confused from the tests because they're constructing their own fields and I think I was using ´IntDocValuesField´ or something. 

I however do think that this might be something to reconsider (at some point in the future though). I guess indexing and querying are much more efficient on integer or longs, and apparently even a [half-float](https://issues.apache.org/jira/browse/LUCENE-7289) was created to mitigate space-intensive indexes on floats. A solution might be to pass the actual indexing-type to the option object of the index function. 

But well, I guess I'd rather wait until I run into an actual issue with query performance before actually implementing such a thing.

**Comment by dmunch** [comment]
I just pushed the last commits which should make this port feature-complete.

If I haven't overseen anything, this port should be fully compatible with existing couchdb-dreyfus instances concerning the documented features. However Lucene 4.x indexes can't be re-used and have to be rebuild, but taken apart that all configuration and queries should stay compatible.

During porting I found two features not mentioned in the official documentation:
- Configuration of the Lucene [Directory](http://lucene.apache.org/core/6_3_0/core/org/apache/lucene/store/Directory.html) in [IndexService.scala](src/main/scala/com/cloudant/clouseau/IndexService.scala#L802)
- Explicit selection of the IndexOptions enum value for string fields in [ClouseauTypeFactory.scala](src/main/scala/com/cloudant/clouseau/ClouseauTypeFactory.scala#L237)

The first one could be easily ported but isn't done yet. The explicit selection of enum types can't really be done in a backward-compatible way since those values are specific to the used Lucene version. 

So far this has been a fun ride, and even though I was new to Lucene and Scala I found it pretty straightforward to understand the code and concepts. It was a perfect playground for learning both and I'd really like to thank the original authors of dreyfus and clouseau for their initial clean work.

I'm not sure which priority Cloudant has to bring those changes into production, loads of components have been updated to make this port work, so the work certainly doesn't stop here and some more thorough testing should be done. 

From my side the work doesn't stop here either, the original motivation was to use the new geo-related features in Lucene > 6.x and the code is now in a state where those can be easily added. I'm going to create a new issue to outline my rough plan to add those, since it certainly also includes changes to dreyfus in order to make geo-related queries work.

Thanks again, hope you all had great holidays and wish you a happy 2017!

**Comment by rnewson** [comment]
Awesome work, thanks for persevering. You are quite right that there's lots of other work to do to bring this to production for Cloudant (seamless upgrade being the hardest one), but that burden does not fall on you. I'm not yet sure whether we could merge this work in yet until there's a bigger plan on our side, but I don't want to see your efforts wasted either.

I assume you're referring to the newDirectory function in that first bullet point. The explanation for that is it gives Cloudant a runtime ability to try other lock factories and directory implementations, and pretty much dates to the original few versions I wrote when I was experimenting with different options to see what performed best in our actual environment. It's possible that none of that is still needed in lucene 6.x.

**Comment by dmunch** [comment]
Ok great. I'm going to use the new version in our internal tests and I'll see how it behaves. In case you're looking at it and need help in clarifying things don't hesitate to ask, I'm happy to help. 

Do you see any issues if I merge the scala-2120 branch in my forked repository into master of my fork? I'd like to keep the branching model clean for my fork, especially since I started implementing https://github.com/cloudant-labs/clouseau/issues/9

**Comment by rnewson** [comment]
that depends on who uses your master branch. If it's just you, then it's fine. The worst case scenario is you'll do a force-push to get back in sync with this repo.

**Comment by brlanier** [comment]
Has there been any progress on these updates? I was using fork of the work done above with success in 2.0.0 and 2.1.0 during limited testing. I have recently migrated to couchdb 2.2.0 and existing full text indexes seem to work but as soon as I try to add a new one it seems something on the clouseau side crashes and the connection is lost. The index shards are not created. 
I was going to try to go back to a default clouseau install to try and troubleshoot, but since it requires java jdk 6 I have run into issues installing that on newer ubuntu. Seems the archives are no longer available for download. 
So just wondering if there are still plans to migrate forward on some level past dependency of jdk 6 or if maybe its just an issue with newer couchdb 2.2.0. I can open a new issue if needed, but this initially seems related and I am using what I feel would be a completely unsupported version of clouseau. 

Here is a snippet of the output from clouseau service if it might be helpful and I'm just missing something more obvious
```
Nov 29 12:24:41 stage mvn[7699]: #011at com.cloudant.clouseau.ClouseauTypeFactory$.addFields(ClouseauTypeFactory.scala:149)
Nov 29 12:24:41 stage mvn[7699]: #011at com.cloudant.clouseau.ClouseauTypeFactory$.$anonfun$readDoc$1(ClouseauTypeFactory.scala:107)
Nov 29 12:24:41 stage mvn[7699]: #011at com.cloudant.clouseau.ClouseauTypeFactory$.$anonfun$readDoc$1$adapted(ClouseauTypeFactory.scala:106)
Nov 29 12:24:41 stage mvn[7699]: #011at com.cloudant.clouseau.ClouseauTypeFactory$.readDoc(ClouseauTypeFactory.scala:106)
Nov 29 12:24:41 stage mvn[7699]: #011at com.cloudant.clouseau.ClouseauTypeFactory$.createType(ClouseauTypeFactory.scala:86)
Nov 29 12:24:41 stage mvn[7699]: 2018-11-29 12:24:41 clouseau.shards/60000000-7fffffff/db_name/5d7f06b0a110d854a9a330304aa32eeb [INFO] Closed with reason: 'noconnection
```

**Comment by pgj** [comment]
Please note that we have moved to a different stack recently in Clouseau 3 which addresses or makes irrelevant most of the suggested changes here.  We are also planning to update Lucene to 10 based on this new foundation, but that is going to managed separately.
